### PR TITLE
fix(parser): resolve Rust mod_item declarations to concrete import edges

### LINKED ISSUE #1000: fix(parser): Rust `mod X;` declarations are not import edges — modules used via qualified paths get zero/undercounted dependents
## Summary

`RustImportExtractor.importNodeTypes = ['use_declaration']` (`packages/parser/src/ast/languages/rust.ts:432`). `mod_item` is not in the list, so a `mod X;` declaration produces no import edge.

That is fine when the parent also writes `use crate::X`, but the idiomatic Rust pattern is `mod X;` at the crate root plus **qualified calls** (`X::func()`) with no `use` at all. In that shape the parser sees no edge whatsoever, and `get_dependents` on the child module returns a confident, wrong answer.

Note `mod_item` **is** already in `RustExportExtractor.exportableTypes` (`rust.ts:117`) — so the node type is recognised on the export side and simply absent on the import side.

## Reproduction (tracked fixture, no setup)

`lien-review-testbed/rust` — 8 files. Parser dependent sets today:

```
src/reporter.rs      deps=[]                        <-- WRONG
src/formatter.rs     deps=["src/reporter.rs"]       <-- INCOMPLETE
src/main.rs          deps=[]                        (correct: crate root)
src/analyzer.rs      deps=["src/cache.rs","src/formatter.rs","src/main.rs","src/reporter.rs"]
src/config.rs        deps=["src/analyzer.rs","src/cache.rs","src/main.rs","src/parser.rs","src/reporter.rs"]
src/error.rs         deps=[... 7 files ...]
src/cache.rs         deps=["src/main.rs","src/reporter.rs"]
src/parser.rs        deps=["src/analyzer.rs","src/reporter.rs"]
```

Ground truth from the source:

```
$ grep -n "reporter" lien-review-testbed/rust/src/main.rs
7:mod reporter;
74:  let cached_report = reporter::report_from_cache(config_path)?;
86:  let report = reporter::generate_and_export(&quick_results, &config)?;
118: let report = reporter::generate_and_export(&results, &config)?;
122: let detailed = reporter::generate_report(&results, &config);
129: reporter::export_report(&results, &report_path)?;
138: let comparison = reporter::compare_results(&window[1], &window[0]);
169: reporter::reanalyze_and_compare(path, &config, &mut result_cache)?;
174: let cached_summary = reporter::report_from_cache(&args[2])?;
236: let diff = reporter::compare_results(&window[1], &window[0]);
247: reporter::export_report(&results, &interim_path)?;
252: reporter::reanalyze_and_compare(path, config, &mut file_cache)?;
```

**`main.rs` calls into `reporter.rs` at 11 sites and `reporter.rs` reports zero dependents.**

`formatter.rs` shows the undercount variant: it gets its single edge from `reporter.rs:7 use crate::formatter;` (a real `use_declaration`), but misses `main.rs`'s 7 qualified `formatter::*` call sites, which are reached only through `mod formatter;`.

What the parser actually extracts from `main.rs` — the 7 `mod` lines are all absent:

```
src/main.rs: imports=["analyzer/AnalysisResult","cache/Cache","config/Config","error/AnalyzerError"]
```

(`main.rs` declares `mod analyzer; mod cache; mod config; mod error; mod formatter; mod parser; mod reporter;`)

## Impact

- `get_dependents` on a `mod`-only module returns `[]` with no caveat — indistinguishable from a genuinely unused file. This is the "confidently wrong" shape, not a visible gap.
- Blast-radius risk scoring underestimates, because dependent count is the primary input.
- Scales with crate size: every module reached by qualified path rather than `use` is invisible.

## Fix sketch — and one trap

Add `mod_item` to `importNodeTypes` and resolve the declared name against a sibling `X.rs` / `X/mod.rs`.

**The trap:** `mod reporter;` yields the bare specifier `reporter`, which is exactly the shape that #928 ("get_dependents fabricates dependency graphs via language-blind bare-module matching") was closed to prevent, and that the #884 guard (`isUnresolvableWholeModuleImport`) exists to reject. A naive fix reintroduces the fabrication bug.

The distinguishing fact is that a Rust `mod X;` is **not** an ambiguous bare-module reference — it is a declaration that resolves deterministically to a specific sibling path relative to the declaring file (`X.rs` or `X/mod.rs`, same directory). So it should be resolved to a concrete path at extraction time rather than emitted as a bare specifier for downstream fuzzy matching. Resolve it, don't guess it.

Worth adding a cross-language assertion here rather than a Rust-only test — the "declaration form is the dependency edge" shape is not unique to Rust.

## Provenance

Found while capturing a pre-refactor dependency baseline for the Phase 3-5 duplication refactor (#994), not by a failing test. `lien-review-testbed/rust` is tracked, so this is reproducible from a clean checkout with no fixture capture.

Distinct from #999 (Rust struct/enum signature metadata).

### PR DESCRIPTION
**Pull request by alfhen**
## Summary

Fixes #1000. `RustImportExtractor.importNodeTypes` only listed
`use_declaration` — `mod_item` was absent, so `mod x;` at the crate root
plus qualified calls (`x::func()`) with no `use` at all produced **zero**
import edges. `get_dependents` on the child module returned a confident,
wrong `[]` (or an undercount when the file happened to also gain one edge
via an unrelated `use crate::x` elsewhere).

Found while baselining #994 (the Phase 3-5 duplication refactor).

**The trap avoided (#928/#884):** `mod reporter;` yields the bare specifier
`reporter`, exactly the shape `isUnresolvableWholeModuleImport` exists to
reject. A Rust `mod X;` is not ambiguous like that, though — it's a
declaration that resolves **deterministically** to a specific sibling module
relative to the declaring file. So the fix resolves `mod` to a
**directory-anchored, extensionless module path at extraction time** (e.g.
`src/reporter` for `src/main.rs`'s `mod reporter;`, honoring the leaf-file
2018+-edition directory split, inline-mod nesting, and a `#[path = "..."]`
override) instead of emitting an unanchored bare specifier for downstream
fuzzy matching to guess at. The extractor itself never appends `.rs` and
doesn't choose between the `x.rs` / `x/mod.rs` on-disk conventions — the
resolved specifier wins via `matchesFile`'s plain **exact-match** strategy
against whichever one is actually on disk (both get extension-stripped by
the existing normalization), no bare-word leniency involved at all.

An inline `mod x { ... }` (has a body — e.g. `#[cfg(test)] mod tests { ... }`)
is a namespace, not an import: it has no separate file and correctly
produces no edge for itself. `collectImportNodes` (`ast/symbols.ts`) now
recurses into such a body so any `use`/nested `mod` declared inside it is
still discovered (previously invisible, since `mod_item` wasn't in the
matched-type set at all).

## Scope

Confined to `packages/parser/src/ast/languages/rust.ts`,
`packages/parser/src/ast/symbols.ts` (the shared `collectImportNodes`
helper — the recursion added there is gated on `child.type === 'mod_item'`,
a no-op for every other language), and their tests, per the no-touch list
for concurrently running Phase 3-5 duplication-refactor agents
(`dependency-analyzer.ts`, `utils/path-matching.ts`,
`insights/chunk-complexity.ts`, `core/insights/complexity-analyzer.ts`,
`review/dependency-graph.ts`, `review/blast-radius.ts` — none touched).

## Before / after: tracked fixture (`lien-review-testbed/rust`)

Measured via `performChunkOnlyIndex` + `findDependents` from `@liendev/parser`
(the same primitives `get_dependents` uses), before and after the fix on the
identical, byte-for-byte tracked fixture — no capture step needed.

| file | before | after |
|---|---|---|
| `src/reporter.rs` | `[]` | **`["src/main.rs"]`** |
| `src/formatter.rs` | `["src/reporter.rs"]` | **`["src/main.rs","src/reporter.rs"]`** |
| `src/parser.rs` | `["src/analyzer.rs","src/reporter.rs"]` | **`["src/analyzer.rs","src/main.rs","src/reporter.rs"]`** |
| `src/main.rs` | `[]` | `[]` (unchanged — crate root, nothing imports it) |
| `src/analyzer.rs` | `["src/cache.rs","src/formatter.rs","src/main.rs","src/reporter.rs"]` | unchanged |
| `src/cache.rs` | `["src/main.rs","src/reporter.rs"]` | unchanged |
| `src/config.rs` | `["src/analyzer.rs","src/cache.rs","src/main.rs","src/parser.rs","src/reporter.rs"]` | unchanged |
| `src/error.rs` | `["src/analyzer.rs","src/cache.rs","src/config.rs","src/formatter.rs","src/main.rs","src/parser.rs","src/reporter.rs"]` | unchanged |

**Note on `src/parser.rs`:** the issue's own reproduction narrative only
called out `reporter.rs` (zero) and `formatter.rs` (undercount) as evidence,
and its "expected after" table lists `parser.rs` unchanged. That's an
oversight in the issue text, not in this fix — `main.rs` declares `mod
parser;` and makes 6 qualified `parser::*` calls (identical shape to
`reporter`/`formatter`), so `parser.rs` **correctly** gains `src/main.rs`
too. Confirmed directly:
```
$ grep -n "parser::" lien-review-testbed/rust/src/main.rs
210:        let input = parser::parse_input(path, config)?;
212:        parser::validate_input(&input)?;
264:    let input = parser::parse_input(path, config)?;
270:    let metadata = parser::parse_metadata(&input);
272:        let (parsed_key, _) = parser::parse_line(...);
311:                let (name, ext) = parser::parse_line(...);
```
So **3 of 8 rows change** (not 2), and the other **5 are byte-identical**
before/after — verified by reverting the two changed source files to `HEAD`,
re-running the same script, and confirming the "before" output matches the
issue's stated ground truth exactly, then reapplying and re-measuring.

## Real-world validation: `dtolnay/anyhow`

Per the "sweep before you fix" ask, also ran against a real, widely-used
crate (shallow-cloned via the same URL the repo's own Rust E2E fixture
uses), not just the 8-file synthetic testbed:

```
lib.rs declares: mod backtrace; mod chain; mod context; mod ensure; mod error;
mod fmt; mod kind; mod macros; mod nightly; mod ptr; mod wrapper;
```

After the fix, every one of those 11 files gains `src/lib.rs` as a
dependent (previously none did):
```
src/backtrace.rs  deps=["src/error.rs","src/lib.rs"]
src/chain.rs      deps=["src/chain.rs","src/error.rs","src/fmt.rs","src/lib.rs"]
src/context.rs    deps=["src/context.rs","src/lib.rs"]
src/ensure.rs     deps=["src/lib.rs"]
src/error.rs      deps=["src/context.rs","src/ensure.rs","src/error.rs","src/fmt.rs",
                        "src/kind.rs","src/lib.rs","tests/test_autotrait.rs",
                        "tests/test_downcast.rs","tests/test_repr.rs"]
src/fmt.rs        deps=["src/lib.rs"]
src/kind.rs       deps=["src/lib.rs"]
src/macros.rs     deps=["src/lib.rs"]
src/nightly.rs    deps=["src/context.rs","src/error.rs","src/lib.rs","src/wrapper.rs"]
src/ptr.rs        deps=["src/error.rs","src/fmt.rs","src/lib.rs"]
src/wrapper.rs    deps=["src/lib.rs"]
src/lib.rs        deps=[]   (crate root, unchanged)
```

This crate also has two REAL inline modules (not present in the synthetic
testbed) that must NOT produce false edges — confirmed clean:
```
context.rs:9   mod ext { use super::*; ... }        <- inline, no ext.rs file exists
fmt.rs:104     mod tests { use super::*; ... }       <- inline, no tests.rs file exists
```
Neither `"ext"` nor `"tests"` appears anywhere in any chunk's `imports` list.

## Sweep: within Rust

Added targeted unit tests (`rust.test.ts`, `symbols.test.ts`) for every
shape called out in the issue:
- `mod x;` at crate root (module-root file) -> same-directory sibling
- `pub mod x;` -> identical resolution (visibility doesn't gate the edge)
- `mod x;` in a **leaf** file (not `main.rs`/`lib.rs`/`mod.rs`) -> resolves
  into a subdirectory named after that file (the 2018+-edition split)
- **Nested inline modules** (`mod outer { mod inner; }`) -> `outer` folds in
  as a directory segment; verified against real `parser.rs`-shape nesting
- **Inline `mod X { ... }`** (has a body, e.g. `mod tests`) -> no edge for
  itself, but `use`/nested `mod` inside it are still discovered
  (`collectImportNodes` recursion)
- **`#[path = "..."]`** override, including with other attributes stacked
  in either order (`#[cfg(test)] #[path = "..."]` and reversed) — resolves
  to the override path relative to the file's own directory, doesn't crash
- No `importerFile` provided -> returns null (can't resolve deterministically,
  matches the existing `self::`/`super::` no-op precedent)

## Sweep: across languages ("declaration form is the edge")

Checked whether any other supported language (TS, JS, Python, Go, Java, C#,
Ruby, Kotlin, Swift, PHP — C isn't a supported AST language here) has a
declaration-style include missing from `importNodeTypes`, the same shape as
Rust's `mod`:

- **Go blank imports** (`import _ "pkg"`) — **confirmed clean**.
  `GoImportExtractor.extractSpecPath`/`extractImportPaths` don't special-case
  the `_`/dot alias at all; the edge is created regardless of alias. Only
  the *symbol name* fallback (`processImportSymbolsList`) treats blank/dot
  specially, which is correct (there's no meaningful symbol name for a blank
  import).
- **Ruby `require`/`require_relative`/`load`/`autoload`** — **confirmed
  clean**. All four are already in `REQUIRE_METHODS` and handled uniformly
  via the `call` node type. Ruby has no separate declaration-only linking
  keyword the way Rust has `mod` — `require`/`require_relative` *are*
  Ruby's only file-linking mechanism, so there's no declaration/use duality
  to miss.
- **PHP `require`/`include`/`require_once`/`include_once`** — **confirmed
  gap, but a different shape**: `PHPImportExtractor.importNodeTypes` is
  `['namespace_use_declaration']` only; literal `require '../foo.php';`
  statements aren't walked as import nodes at all. This is not the
  declaration-vs-use duality #1000 is about (PHP has no `mod`-like
  declarative keyword) — it's a separate, pre-existing "one whole import
  mechanism isn't implemented for PHP" gap, worth its own follow-up issue.
  Not fixed here (out of scope; not requested by #1000, no existing test
  coverage to regress, real design work needed for path resolution incl.
  dynamic requires).
- **C# `partial` classes / Java & Kotlin same-package implicit access** —
  a *different* shape (no import statement is needed at all for same-
  namespace/package files, not a missed declaration keyword). Already
  tracked/being addressed separately: C# namespace-scoped dependent recovery
  shipped in #971, and Java/Go already have dedicated same-package/directory
  test-pairing logic (`java-same-package-tests.test.ts`,
  `go-same-directory-tests.test.ts`) for this exact concern. Not duplicated
  here.
- **Swift module declarations** — already the `wholeModuleImports: true`
  gated case (#884); architecturally a different, already-solved shape
  (Swift's `import Foundation` never names a specific file at all).
- **C `#include`** — not applicable; C isn't a currently supported AST
  language in this parser.

**Verdict:** Rust's `mod` is the only instance of this specific shape
among currently-supported languages. No cross-language test table was
added since the shape doesn't recur elsewhere yet (per the repo policy,
a table is for asserting a *recurring* cross-language invariant — this one
doesn't recur, so a Rust-scoped fix + Rust-scoped tests is the correct-sized
artifact).

## Gates

```
npm run format:check   -> pass
npm run lint            -> 0 errors (37 pre-existing warnings, unrelated files)
npm run typecheck       -> 0 errors
npm run build           -> success
npm test                -> exit 0, all packages pass (parser 1619/1619,
                             core 406/406, review 1723/1723, cli 1525/1525,
                             action 41/41, parser-native 1/1 — 5315/5315 total)
lien delta               -> exit 0 (3 "worsened"/9 "new" entries, all well under threshold —
                             see below)
```

`lien delta` output (informational only — nothing crossed a threshold):
```
packages/parser/src/ast/languages/rust.ts
  ⚠ worsened  RustImportExtractor.extractImportPath      cognitive 2 → 3
  ⚠ worsened  RustImportExtractor.processImportSymbols   cognitive 1 → 3
  · new       basenameNoExt                    cyclomatic – → 1
  · new       rustModOwningDirectory           cognitive  – → 1
  · new       collectInlineAncestorModNames    cognitive  – → 6
  · new       isInlineModDeclaration           cyclomatic – → 1
  · new       findPathAttributeOverride        cognitive  – → 7
  · new       extractPathAttributeValue        cognitive  – → 2
  · new       extractModImportPath             cognitive  – → 4

packages/parser/src/ast/symbols.ts
  ⚠ worsened  collectImportNodes               time 11m → 13m
  · new       collectNestedModImportNodes      cognitive – → 2
```
(`collectImportNodes` was split into a helper mid-implementation specifically
because the first version pushed its own cognitive complexity 11→18, over
the threshold of 15 — the version shipped here stays under it.)

Also run (dogfooded, real dependency checks — not synthetic):
```
npm run test:e2e:rust -w packages/cli   -> 14 passed, 140 skipped (network-gated), 0 failed
npm run test:e2e:go -w packages/cli     -> 14 passed, 140 skipped (network-gated), 0 failed
```

## Notes

- `parser` is published — changeset added (`patch`, matching this repo's
  precedent for parser-only bugfixes).
- No files from the Phase 3-5/4/5 duplication-refactor no-touch list were
  edited.

---

### Lien Review

➡️ **Stable** - 1 pre-existing issue in touched files (none introduced).
| Metric | Violations | Change |
|--------|:----------:|:------:|
| 🧠 mental load | 1 | +0 |

> [!NOTE]
> **Low Risk**
>
> This PR fixes a real bug in Rust import extraction by adding `mod_item` to the matched import-node types and resolving file-backed `mod x;` declarations to directory-anchored, extensionless module paths. The change is additive and confined to the Rust extractor and the shared `collectImportNodes` helper. Callers in `chunker.ts` receive additional, more accurate import edges for Rust modules; no existing caller is broken because the new `mod_item` handling returns `null` safely when `importerFile` is absent and skips inline `mod` bodies. The implementation correctly distinguishes module-root files (`main.rs`/`lib.rs`/`mod.rs`) from leaf files, honors inline-mod nesting and `#[path = "..."]` overrides, and recurses into inline mod bodies without fabricating edges for the inline module itself. Extensive unit and integration tests cover the new behavior, including the boundary cases called out in the issue.
>
> - Added `mod_item` to `RustImportExtractor.importNodeTypes` and implemented deterministic path resolution for file-backed `mod x;` declarations.
> - Added `collectNestedModImportNodes` so `use` and nested `mod` declarations inside inline `mod` blocks are still discovered without creating a false edge for the inline block.
> - Added targeted tests for module-root files, leaf files, inline nesting, `#[path]` overrides, and inline mod blocks.

✅ *Trust: **Delivered** — The review ran to completion within budget.*

<sup>Reviewed by [Lien Review](https://lien.dev) for commit 72b046e. Updates automatically on new commits.</sup>

### CODE REVIEW

File: packages/parser/src/ast/languages/rust.ts [hunk 1]
```diff
@@ -210,6 +210,142 @@ function joinFromDir(dir: string, rest: string): string {
...
+function basenameNoExt(p: string): string {
+  const base = p.slice(p.lastIndexOf('/') + 1);
+  return base.replace(/\.rs$/, '');
+}
+
+/**
+ * The directory a FILE-BACKED `mod_item` (`mod x;`, no body) resolves its
+ * sibling file against — Rust's file-to-module convention (#1000):
+ *
+ * - A module-root file (`mod.rs`/`lib.rs`/`main.rs`) owns its OWN directory:
```
[the remark this answers is not in the document]
Comment (alfhen) [reply packages/parser/src/ast/languages/rust.ts:235]: Fixed in 72b046e9 — reworded to describe the actual extensionless, directory-anchored specifier (`src/reporter`) rather than a `.rs`-suffixed file path. The follow-up Lien Review run against that commit confirms no remaining doc-truth findings.

File: .changeset/rust-mod-item-import-edges.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+---
+
+Fix Rust `mod X;` declarations producing no import edge (#1000).
+
+`RustImportExtractor.importNodeTypes` only listed `use_declaration` —
+`mod_item` was absent, so the idiomatic Rust pattern of `mod x;` at the
+crate root plus qualified calls (`x::func()`) with no `use` at all produced
+zero import edges. `get_dependents` on the child module returned a
+confident, wrong `[]` (or an undercount when the file also happened to gain
+one edge via an unrelated `use crate::x` elsewhere).
```
[the remark this answers is not in the document]
Comment (alfhen) [reply .changeset/rust-mod-item-import-edges.md:12]: Fixed in 72b046e9 — changeset now says the extractor resolves to the extensionless module path, with downstream matching mapping it to `x.rs` or `x/mod.rs`. Confirmed clean on the follow-up Lien Review run against that commit.
